clippy_lints/src/non_copy_const.rs

```diff
@@ -110,10 +112,15 @@ impl Source {
 }
 
 fn verify_ty_bound<'tcx>(cx: &LateContext<'tcx>, ty: Ty<'tcx>, source: Source) {
-    if ty.is_freeze(cx.tcx.at(DUMMY_SP), cx.param_env) || is_copy(cx, ty) {
-        // An `UnsafeCell` is `!Copy`, and an `UnsafeCell` is also the only type which
-        // is `!Freeze`, thus if our type is `Copy` we can be sure it must be `Freeze`
-        // as well.
+    // Ignore types whose layout is unknown since `is_freeze` reports every generic types as `!Freeze`,
+    // making it indistinguishable from `UnsafeCell`. i.e. it isn't a tool to prove a type is
+    // 'unfrozen'. However, this code causes a false negative in which
+    // a type contains a layout-unknown type, but also a unsafe cell like `const CELL: Cell<T>`.
```

Comment:
Can you add a test for this, with the remark, that this is a FN? (Or maybe I just missed the test)

Reply:
You're right; I didn't write the test for it and I'm happy to do so. However, I probably hesitated to add more things to this already complicated test since just opening the test file is enough to make my mind go blank. The current state of the test is a by-product of my attempt to retain the original form as much as possible. And, I've thought about it, how about splitting the traits and giving these some self-descriptive name? For example:

```rust
trait ConcreteDefinitions {
    const ATOMIC: AtomicUsize;
    const INTEGER: u64;
}

trait AssocTypes {
    type TypeToBeFrozen;
    type TypeToBeUnfrozen;

    const CONST_TO_BE_FROZEN: TypeToBeFrozen;
    const CONST_TO_BE_UNFROZEN: TypeToBeUnfrozen;
}

trait BothOfCellAndGeneric<T> {
    const DIRECT: Cell<T>;
    const INDIRECT: Cell<*const T>;
}
```

...and so on (plus impls of these). As a downside, it requires rewriting the test, makes the diff useless.

Reply:
I also struggled to understand the tests and had to go back and forth to understand what's going on. So splitting it up in multiple traits or otherwise improving the tests would be great! You can also create a dir for these tests and split them up in multiple files, if you think that would help. 

So I'm happy how the tests are currently (+the FN test), but would be happier with a better/easier to understand test structure.

Reply:
Thank you and I'm glad to hear it's OK for you to review a quite heavily restructured test!

And, you were *right*; it've caught a little false positive where `Cell<*const T>` gets linted at the trait impl, in addition to the trait def.

Reply:
> `Cell<*const T>` gets linted at the trait impl, in addition to the trait def.

Did you already fix that? I don't see this behavior in the `.stderr` file.

Reply:
Yes, I've bundled it with the test rewrite. The part of that which changes `has_type_flags` to `layout_of` in `non_copy_const.rs`.